### ISSUE TITLE
“T[][] is not assignable to Iterable<[T, T]>” while it is (at compile time)

### ISSUE THREAD

**Issue by msklvsk** [Microsoft/TypeScript#7799]
**TypeScript Version:** 

1.8.9

**Code**

```ts
new Map<number, number>([1, 2].map(x => [x, x]));

// Error:
// Argument of type 'number[][]' is not assignable to parameter of type 'Iterable<[number, number]>'…

```

**Expected behavior:** types are inferred correctly, no error shown.
**Actual behavior:** type error.

**Comment by FranklinWhale** [comment]
I think this issue is related to #6574

**Comment by msklvsk** [comment]
Thanks, yes, same problem, a dupe. Should I close this one?

**Comment by Arnavion** [comment]
It would be nice if contextual types allowed to infer an array as a tuple, to support such an idiom.

@FranklinWhale I don't think #6574 is related. If #6574 was fixed, the inferred type would be `Iterable<number[]>` which would still be a problem. This issue is about inferring `[x, x]` as a `[number, number]` instead of `number[]`.

**Comment by FranklinWhale** [comment]
I don't mean that this issue is a duplicate of #6574, but they are both related to array to tuple inference. I am sorry if any confusion has been caused.

To be clear:
In #6474, `T[]` may be a tuple of T.
In this issue, `T[][]` may be an array of tuples of T.

**Comment by Arnavion** [comment]
#6574 is about inferring `Tuple<T>.map<U>()`'s result as a `Tuple<U>` instead of `Array<U>`. It says nothing about inferring array literals as tuples (notice that it starts with explicit annotation for the tuple), whereas this issue is about inferring array literals as tuples. I.e., if `[x, x]` was inferred as a `[number, number]` instead of `number[]`

Hence my point that they're not related.

**Comment by FranklinWhale** [comment]
That's not really the case. In @elbarzil's code sample in #6574, `[1,2]` in `[1,2].map(x=>x+1)` is actually a `number[]`, and the `map` function also returns a `number[]`. It is actually the same as assigning an array created from an array literal to a varible declared as a tuple. Therefore, #6574 will be fixed when any array may be inferred as tuple when there is an matching tuple declaration.

**Comment by FranklinWhale** [comment]
See also #3369
